ext/vscode/src/utils/authServer.ts

```diff
@@ -0,0 +1,90 @@
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import * as http from 'http';
+import { randomBytes } from 'crypto';
+import { TokenCredential } from '@azure/core-auth';
+import { isNotSignedInError } from './vsCodeAuthenticationCredential';
+
+// startAuthServer creates a locally running server that will respond to Azure Dev CLI authentication requests and
+// starts listening for requests.  Requests must be authenticated with a key that is returned from this function.
+// The provided credential is used to fetch tokens for auth requests.
+export function startAuthServer(credential: TokenCredential): Promise<{ server: http.Server, key: string }> {
+    const key = randomBytes(32).toString('hex');
+
+    const server = http.createServer(async (req, res) => {
+        if (req.headers['content-type'] !== 'application/json' || req.method !== 'POST' || req.url !== '/token?api-version=2023-07-12-preview') {
+            res.writeHead(400).end();
+            return;
+        }
+
+        if (req.headers['authorization'] !== `Bearer ${key}`) {
+            res.writeHead(401).end();
+            return;
+        }
+
+        const body: Buffer[] = [];
+        req.on('data', (chunk) => { body.push(chunk); });
+        req.on('end', async () => {
+            const reqBody = JSON.parse(Buffer.concat(body).toString());
+            if (typeof reqBody !== 'object') {
+                res.writeHead(400).end();
+                return;
+            }
+
+            if (!reqBody.scopes ||
+                !Array.isArray(reqBody.scopes) ||
+                reqBody.scopes.length === 0 ||
+                reqBody.scopes.some((a: unknown) => typeof a !== 'string') ||
+                reqBody.tenantId && typeof reqBody.tenantId !== 'string') {
+                res.writeHead(400).end();
+                return;
+            }
+
+            res.setHeader('Content-Type', 'application/json');
+
+            try {
+                const token = await credential.getToken(reqBody.scopes, {
+                    tenantId: reqBody.tenantId,
+                });
+
+                if (!token) {
+                    res.writeHead(500).end();
+                    return;
+                }
+
+                res.writeHead(200).end(
+                    JSON.stringify({
+                        status: 'success',
+                        token: token.token,
+                        expiresOn: new Date(token.expiresOnTimestamp).toISOString(),
+                    })
+                );
+            } catch (e: unknown) {
+                let message = (e instanceof Error) ? e.message : 'unknown error';
+                let code = "GetTokenError";
+
+                if (isNotSignedInError(e)) {
+                    code = "NotSignedInError";
+                    message = 'You are not signed in to an Azure account. Please sign in.';
+                }
+
+                res.writeHead(200).end(
```

Comment:
Shouldn't it be a 5xx error here?

Reply:
Reworked this a little - the intention is that if GetToken() fails we return 200 OK but with a /failure/ status code, so we have some way to shuttle the failure to get a token back to the other side.  If other things fail, we do want to 500.